File: 19_removing_nth_node_from_the_end.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int k) {
        if(!head || !head->next)return NULL;
        ListNode dummy(0), *cur = &dummy;
        cur->next = head;
        ListNode* former = cur, *latter = cur;
        while(k){
            latter = latter->next;
            k--;
        }
        while(latter && latter->next){
            former = former->next;
            latter = latter->next;
        }
        former->next = former->next->next; // ? former->next->next : NULL;
        return dummy.next;
    }
};

File: 19_removing_nth_node_from_the_end.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dumy = ListNode()
        dumy.next = head
        first = dumy
        second = dumy
        for i in range(0,n+1):
            first = first.next

        while first:
            first = first.next
            second = second.next

        second.next = second.next.next
        return dumy.next
    

# Two pass solution TC: O(2*N), SC: O(1)
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        size = 0
        temp1 = head
        while temp1:
            temp1 = temp1.next
            size += 1
        remove_index = size - n + 1
        temp1 = head
        cur_index=1
        if remove_index == 1:
            return head.next
        while temp1:
            if remove_index == cur_index+1:
                temp1.next=temp1.next.next if temp1.next else None
                break
            temp1 = temp1.next
            cur_index+=1

        return head